### ISSUE TITLE
vector.make crashes with NotImplementedError in envs with custom spaces, and Box crashes with None in shape

### ISSUE THREAD

**Issue by bionicles** [openai/gym#1972]
To use RL for datasets, I made a String space, and unfortunately, this crashes when you attempt to make vector env with your custom env with custom spaces:

```
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/__init__.py", line 61, in make
    return AsyncVectorEnv(env_fns) if asynchronous else SyncVectorEnv(env_fns)
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/async_vector_env.py", line 88, in __init__
    observation_space=observation_space, action_space=action_space)
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/vector_env.py", line 34, in __init__
    self.observation_space = batch_space(observation_space, n=num_envs)
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/utils/spaces.py", line 41, in batch_space
    return batch_space_dict(space, n=n)
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/utils/spaces.py", line 70, in batch_space_dict
    for (key, subspace) in space.spaces.items()]))
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/utils/spaces.py", line 70, in <listcomp>
    for (key, subspace) in space.spaces.items()]))
  File "/home/bion/miniconda3/lib/python3.7/site-packages/gym/vector/utils/spaces.py", line 43, in batch_space
    raise NotImplementedError(
```

The issue is, it's not clear how to make a String space out of the regular spaces in gym, because strings are variable length; we could pad them, but padding is wasteful

Can we add a simple escape hatch to just return a tuple of any-type stuff if the space isn't in the list of supported batch spaces?

Here's the line: https://github.com/openai/gym/blob/809a8ad21ba66b2a17987005a04be3881808bf37/gym/vector/utils/spaces.py#L43

Side note: I'm pretty sure we can drop Binary, MultiBinary, MultiDiscrete, and maybe Discrete, if you guys want to reduce the size of the codebase for maintenance simplification, everything is ultimately just a Box, and Box can model every single one of those other spaces. 

If Box supported None or -1 'ragged' dimensions, then it could handle strings, molecules, audio, etc ... basically all we'd have to do, is instead of erroring in Box.__init__, error in Box.sample, if there is a None or -1 in self.shape. This could be considered a separate issue but, if you try to make a ragged box, it crashes on the call to broadcast low / high values to the shape of the box:

https://github.com/openai/gym/blob/809a8ad21ba66b2a17987005a04be3881808bf37/gym/spaces/box.py#L43

**Comment by bionicles** [comment]
let me know if i should split this into 2 issues

**Comment by bionicles** [comment]
TLDR goals to fix:
1. Allow vectorization of envs with custom spaces
2. Allow Box space to have "ragged" shape (None or -1) for a dimension
